fastly/paginator.go

```diff
@@ -1,31 +1,135 @@
 package fastly
 
-// TODO: In go 1.18 (Feb 2022) use generics to reduce the duplicated code.
+import (
+	"net/url"
+	"strconv"
 
-// PaginatorACLEntries represents a paginator.
-type PaginatorACLEntries interface {
-	HasNext() bool
-	Remaining() int
-	GetNext() ([]*ACLEntry, error)
+	"github.com/peterhellberg/link"
+)
+
+// PaginatorKVStoreEntries represents a paginator for KV Store entries.
+type PaginatorKVStoreEntries interface {
+	Next() bool
+	Keys() []string
+	Err() error
 }
 
-// PaginatorDictionaryItems represents a paginator.
-type PaginatorDictionaryItems interface {
-	HasNext() bool
-	Remaining() int
-	GetNext() ([]*DictionaryItem, error)
+// NOTE: We can't identify the underlying type of the type parameter T.
+// This is because we don't assign it to any of the defined function parameters.
+// If we did, then we could do this: https://go.dev/play/p/dfTMGjaSSAX.
+//
+// This means we have to have the caller pass the API path.
+// To make things easier we expose functions for constructing each path.
+// e.g. fastly.ServicePath(), fastly.ACLEntriesPath() etc.
+func NewPaginator[T any](client *Client, input *ListInput, path string) *ListPaginator[T] {
```

Comment:
I mentioned this elsewhere but since the construction of a paginator is an implementation detail and all the uses of it are in the same package, you could unexport this.

Reply:
OK have made this private.